service/spawn.go

```diff
@@ -469,3 +467,100 @@ func (uis *UIServer) modifySpawnHost(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+func (uis *UIServer) modifyVolume(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	updateParams := restModel.APIVolumeModify{}
+	if err := utility.ReadJSON(util.NewRequestReader(r), &updateParams); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	volumeID := gimlet.GetVars(r)["volume_id"]
+	vol, err := host.FindVolumeByID(volumeID)
+	if err != nil {
+		uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "error finding volume '%s'", volumeID))
+		return
+	}
+	if vol == nil {
+		uis.LoggedError(w, r, http.StatusNotFound, errors.Wrapf(err, "no volume '%s' exists", volumeID))
+		return
+	}
+
+	u := MustHaveUser(r)
+	if u.Username() != vol.CreatedBy {
+		uis.LoggedError(w, r, http.StatusUnauthorized, errors.New("not authorized to modify this volume"))
+		return
+	}
+
+	if updateParams.Action == nil {
+		uis.LoggedError(w, r, http.StatusBadRequest, errors.New("no action specified"))
+		return
+	}
+
+	mgrOpts := cloud.ManagerOpts{
+		Provider: evergreen.ProviderNameEc2OnDemand,
+		Region:   cloud.AztoRegion(vol.AvailabilityZone),
+	}
+	mgr, err := cloud.GetManager(ctx, uis.env, mgrOpts)
+	if err != nil {
+		uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "can't get manager for volume '%s'", vol.ID))
+		return
+	}
+
+	// take the specified action
+	switch *updateParams.Action {
+	case VolumeRename:
+		uis.LoggedError(w, r, http.StatusUnauthorized, errors.Wrapf(vol.SetDisplayName(*updateParams.NewName), "can't set display name of '%s' to '%s'", vol.ID, *updateParams.NewName))
+
+	case VolumeExtendExpiration:
+		//TODO
+
+	case VolumeAttach:
+		if updateParams.HostID == nil || *updateParams.HostID == "" {
+			uis.LoggedError(w, r, http.StatusBadRequest, errors.New("must specify host id"))
+			return
+		}
+		var h *host.Host
+		h, err = host.FindOneId(*updateParams.HostID)
+		if err != nil {
+			uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "can't get host '%s'", vol.Host))
+			return
+		}
+		if h == nil {
+			uis.LoggedError(w, r, http.StatusBadRequest, errors.Errorf("host '%s' does not exist", *updateParams.HostID))
+			return
+		}
+
+		if err = mgr.AttachVolume(ctx, h, &host.VolumeAttachment{VolumeID: vol.ID}); err != nil {
+			uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "can't attach volume '%s'", vol.ID))
+			return
+		}
+
+	case VolumeDetach:
+		if vol.Host == "" {
+			uis.LoggedError(w, r, http.StatusBadRequest, errors.Wrapf(err, "volume '%s' is not attached", vol.ID))
+			return
+		}
+		h, err := host.FindOneId(vol.Host)
+		if err != nil {
+			uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "can't get host '%s' for volume '%s'", vol.Host, vol.ID))
+			return
+		}
+		if h == nil {
+			uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrapf(err, "host '%s' for volume '%s' doesn't exist", vol.Host, vol.ID))
```

Comment:
If the host doesn't exist, should we remove vol.Host?

Reply:
done